manta-crypto/src/algebra.rs

```diff
@@ -113,6 +141,111 @@ impl<G, const N: usize> PrecomputedBaseTable<G, N> {
     }
 }
 
+/// Group Element Table for Windowed Point Multiplication
+#[derive(Clone, Debug, Default, Eq, Hash, PartialEq)]
+pub struct Window<G> {
+    /// Multiplication Table
+    table: Vec<G>,
+}
+
+impl<G> Window<G> {
+    /// Creates a new [`Window`] from `table` without checking its correctness.
+    #[inline]
+    pub fn new_unchecked(table: Vec<G>) -> Self {
+        Self { table }
+    }
+
+    /// Creates a new [`Window`] table by repeatedly adding `point` to itself to
+    /// support a table with windowed multiplication with `window_size`.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if `window_size` is less than `1`.
+    ///
+    /// # Implementation Note
+    ///
+    /// Windowed multiplication consists of P/n rounds, where P is the number of bits
+    /// of a group element and n is the window size. Creating the table costs `2^n - 2`
+    /// additions in the group, and each round involves `2^n - 1` table look-ups,
+    /// `n` doublings and 1 addition. In R1CS, the number of constraints becomes
+    /// `6(2^n-2) + P/n [2(2^n-1) + 5n + 6]`, making the optimal window size `n=2`.
+    #[inline]
+    pub fn new<COM>(window_size: usize, point: G, compiler: &mut COM) -> Self
+    where
+        G: Clone + Group<COM> + Zero<COM>,
+    {
+        assert!(window_size > 0, "Window size must be at least 1.");
+        let table_length = 2usize.pow(window_size as u32);
+        let mut table = Vec::with_capacity(table_length);
+        table.push(G::zero(compiler));
+        table.push(point.clone());
+        for _ in 2..table_length {
+            table.push(table.last().unwrap().add(&point, compiler));
+        }
+        Self::new_unchecked(table)
+    }
+
+    /// Returns the window size.
+    #[inline]
+    pub fn window_size(&self) -> usize {
+        self.table.len().trailing_zeros() as usize
+    }
+
+    /// Returns a shared reference to the multiplication table.
+    #[inline]
+    pub fn table(&self) -> &[G] {
+        &self.table
+    }
+
+    /// Returns the multiplication table, dropping `self`.
+    #[inline]
+    pub fn into_inner(self) -> Vec<G> {
+        self.table
+    }
+
+    /// Doubles `result`, `window_size`-many times and then adds the element from `table` corresponding to `chunk`.
+    #[inline]
+    fn scalar_mul_round<COM>(
+        window_size: usize,
+        table: &[G],
+        chunk: &[&Bool<COM>],
+        result: &mut G,
+        compiler: &mut COM,
+    ) where
+        G: Clone + ConditionalSelect<COM> + Group<COM>,
+        COM: Has<bool>,
+    {
+        let chunk = chunk.iter().copied().rev();
```

Comment:
On the contrary, `chunk` (the input of `scalar_mul_round`) goes from most to least significant. We take `rev()` here because `select_from_table` starts from the first bit and chooses the first/second element of each pair of the table, we want the least significant bit to do this first distinction. If you look at the test, you'll see that I use `to_bits_be`, so everything works in the big-endian representation. I'll add an implementation note for the `scalar_mul` method specifying that `bits` must be big-endian for it to be compatible with `select_from_table` and `Window::new`.